src/components/application_manager/rpc_plugins/sdl_rpc_plugin/src/commands/mobile/perform_interaction_request.cc

```diff
@@ -1063,13 +1104,111 @@ void PerformInteractionRequest::SendBothModeResponse(
       msg_param.empty() ? NULL : &msg_param;
   std::string info = app_mngr::commands::MergeInfos(
       ui_perform_info, ui_info_, vr_perform_info, vr_info_);
+
   DisablePerformInteraction();
+
   SendResponse(result,
                perform_interaction_result_code,
                info.empty() ? NULL : info.c_str(),
                response_params);
 }
 
+mobile_apis::Result::eType
+PerformInteractionRequest::PrepareResultCodeForResponse(
+    const app_mngr::commands::ResponseInfo& ui_response,
+    const app_mngr::commands::ResponseInfo& vr_response) {
+  LOG4CXX_DEBUG(
+      logger_, "InteractionMode = " << static_cast<int32_t>(interaction_mode_));
+
+  auto result = mobile_api::Result::INVALID_ENUM;
+  switch (interaction_mode_) {
+    case mobile_apis::InteractionMode::VR_ONLY: {
+      result = MessageHelper::HMIToMobileResult(vr_result_code_);
+      break;
+    }
+
+    case mobile_apis::InteractionMode::MANUAL_ONLY: {
+      result = MessageHelper::HMIToMobileResult(ui_result_code_);
+      break;
+    }
+
+    case mobile_apis::InteractionMode::BOTH: {
+      if (IsVRPerformInteractionResponseSuccessfulInBothMode()) {
+        return MessageHelper::HMIToMobileResult(vr_result_code_);
+      } else {
+        return MessageHelper::HMIToMobileResult(ui_result_code_);
+      }
+      break;
+    }
+
+    default:
+      result = CommandRequestImpl::PrepareResultCodeForResponse(ui_response,
+                                                                vr_response);
+  }
+
+  return result;
+}
+
+bool PerformInteractionRequest::PrepareResultForMobileResponse(
+    app_mngr::commands::ResponseInfo& ui_response,
+    app_mngr::commands::ResponseInfo& vr_response) const {
+  bool result = false;
+  switch (interaction_mode_) {
+    case mobile_apis::InteractionMode::VR_ONLY:
+      result = vr_response.is_ok;
+      break;
+    case mobile_apis::InteractionMode::MANUAL_ONLY:
+      result = ui_response.is_ok;
+      break;
+    case mobile_apis::InteractionMode::BOTH:
+      result = (vr_response.is_ok || ui_response.is_ok);
+      break;
+    default:
+      result = CommandRequestImpl::PrepareResultForMobileResponse(ui_response,
+                                                                  vr_response);
+  }
+
+  return result;
+}
+
+bool PerformInteractionRequest::
+    IsVRPerformInteractionResponseSuccessfulInBothMode() {
+  using namespace mobile_apis;
+  app_mngr::commands::ResponseInfo vr_perform_info(
+      vr_result_code_, HmiInterfaces::HMI_INTERFACE_VR, application_manager_);
+  return (vr_perform_info.is_ok && InteractionMode::BOTH == interaction_mode_);
+}
+
+void PerformInteractionRequest::SetChoiceIdToResponseMsgParams(
+    ns_smart_device_link::ns_smart_objects::SmartObject& msg_param) {
+  LOG4CXX_AUTO_TRACE(logger_);
+
+  std::int32_t choice_id = INVALID_CHOICE_ID;
```

Comment:
this variable is unnecessary, you can set the choice id right on the smart object